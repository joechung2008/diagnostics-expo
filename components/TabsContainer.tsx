import { useApp } from '@/components/AppContext';
import { useTheme } from '@/components/ThemeContext';
import { Ionicons } from '@expo/vector-icons';
import { Tabs, usePathname } from 'expo-router';
import React from 'react';
import { StyleSheet, View } from 'react-native';

export default function TabsContainer() {
  const pathname = usePathname();
  const { colors } = useTheme();
  const { currentId } = useApp();

  return (
    <View style={styles.tabsContainer}>
      <Tabs
        screenOptions={{
          headerShown: false, // Header is in Header component
          tabBarActiveTintColor: colors.primary,
          tabBarInactiveTintColor: colors.textSecondary,
          tabBarStyle: {
            backgroundColor: colors.surface,
            borderTopColor: colors.border,
          },
        }}
      >
        <Tabs.Screen
          name="index"
          options={{
            title: 'Extensions',
            tabBarIcon: ({ color, size }) => (
              <Ionicons name="list" size={size} color={color} />
            ),
            href: pathname !== '/' ? '/' : null,
          }}
        />
        <Tabs.Screen
          name="extensions/[id]"
          options={{
            title: 'Extension',
            tabBarIcon: ({ color, size }) => (
              <Ionicons name="document-text" size={size} color={color} />
            ),
            href:
              pathname === '/' && currentId ? `/extensions/${currentId}` : null,
          }}
        />
        <Tabs.Screen
          name="build"
          options={{
            title: 'Build Info',
            tabBarIcon: ({ color, size }) => (
              <Ionicons name="build" size={size} color={color} />
            ),
          }}
        />
        <Tabs.Screen
          name="server"
          options={{
            title: 'Server Info',
            tabBarIcon: ({ color, size }) => (
              <Ionicons name="server" size={size} color={color} />
            ),
          }}
        />
      </Tabs>
    </View>
  );
}

const styles = StyleSheet.create({
  tabsContainer: {
    flex: 1,
  },
});
