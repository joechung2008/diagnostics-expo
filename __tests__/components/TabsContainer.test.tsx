import TabsContainer from '@/components/TabsContainer';
import { render } from '@testing-library/react-native';
import React from 'react';

// Mock ThemeContext
jest.mock('@/components/ThemeContext', () => ({
  useTheme: () => ({
    colors: {
      primary: '#007AFF',
      textSecondary: '#8E8E93',
      surface: '#F2F2F7',
      border: '#C6C6C8',
    },
  }),
}));

// Create a mock for useApp
const mockUseApp = jest.fn();

// Mock AppContext
jest.mock('@/components/AppContext', () => ({
  useApp: () => mockUseApp(),
}));

// Mock expo-router
const mockUsePathname = jest.fn();
jest.mock('expo-router', () => {
  const MockTabs = ({ children, ...props }: any) => {
    const createElement = require('react').createElement;
    return createElement('Tabs', props, children);
  };

  MockTabs.Screen = ({ ...props }: any) => {
    const createElement = require('react').createElement;
    return createElement('TabsScreen', props);
  };

  return {
    Tabs: MockTabs,
    usePathname: () => mockUsePathname(),
  };
});

// Mock expo/vector-icons
jest.mock('@expo/vector-icons', () => ({
  Ionicons: 'MockIonicons',
}));

describe('TabsContainer', () => {
  beforeEach(() => {
    mockUseApp.mockReturnValue({
      currentId: 'test-extension-id',
    });
    mockUsePathname.mockReturnValue('/');
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders with default state', () => {
    const { toJSON } = render(<TabsContainer />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders with current extension ID', () => {
    mockUseApp.mockReturnValue({
      currentId: 'my-extension-123',
    });

    const { toJSON } = render(<TabsContainer />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders when on extension detail page', () => {
    mockUsePathname.mockReturnValue('/extensions/test-id');

    const { toJSON } = render(<TabsContainer />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders when no current extension ID is set', () => {
    mockUseApp.mockReturnValue({
      currentId: null,
    });

    const { toJSON } = render(<TabsContainer />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('applies theme colors correctly', () => {
    const { toJSON } = render(<TabsContainer />);

    // The snapshot will capture the theme colors applied to the tab bar
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders all tab screens with correct titles and icons', () => {
    const { toJSON } = render(<TabsContainer />);

    // This test verifies that all 4 tabs (Extensions, Extension, Build Info, Server Info) are rendered
    expect(toJSON()).toMatchSnapshot();
  });
});
