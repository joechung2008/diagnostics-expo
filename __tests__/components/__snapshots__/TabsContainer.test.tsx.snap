// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabsContainer applies theme colors correctly 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <Tabs
    screenOptions={
      {
        "headerShown": false,
        "tabBarActiveTintColor": "#007AFF",
        "tabBarInactiveTintColor": "#8E8E93",
        "tabBarStyle": {
          "backgroundColor": "#F2F2F7",
          "borderTopColor": "#C6C6C8",
        },
      }
    }
  >
    <TabsScreen
      name="index"
    />
    <TabsScreen
      name="extensions/[id]"
    />
    <TabsScreen
      name="build"
    />
    <TabsScreen
      name="server"
    />
  </Tabs>
</View>
`;

exports[`TabsContainer renders all tab screens with correct titles and icons 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <Tabs
    screenOptions={
      {
        "headerShown": false,
        "tabBarActiveTintColor": "#007AFF",
        "tabBarInactiveTintColor": "#8E8E93",
        "tabBarStyle": {
          "backgroundColor": "#F2F2F7",
          "borderTopColor": "#C6C6C8",
        },
      }
    }
  >
    <TabsScreen
      name="index"
    />
    <TabsScreen
      name="extensions/[id]"
    />
    <TabsScreen
      name="build"
    />
    <TabsScreen
      name="server"
    />
  </Tabs>
</View>
`;

exports[`TabsContainer renders when no current extension ID is set 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <Tabs
    screenOptions={
      {
        "headerShown": false,
        "tabBarActiveTintColor": "#007AFF",
        "tabBarInactiveTintColor": "#8E8E93",
        "tabBarStyle": {
          "backgroundColor": "#F2F2F7",
          "borderTopColor": "#C6C6C8",
        },
      }
    }
  >
    <TabsScreen
      name="index"
    />
    <TabsScreen
      name="extensions/[id]"
    />
    <TabsScreen
      name="build"
    />
    <TabsScreen
      name="server"
    />
  </Tabs>
</View>
`;

exports[`TabsContainer renders when on extension detail page 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <Tabs
    screenOptions={
      {
        "headerShown": false,
        "tabBarActiveTintColor": "#007AFF",
        "tabBarInactiveTintColor": "#8E8E93",
        "tabBarStyle": {
          "backgroundColor": "#F2F2F7",
          "borderTopColor": "#C6C6C8",
        },
      }
    }
  >
    <TabsScreen
      name="index"
    />
    <TabsScreen
      name="extensions/[id]"
    />
    <TabsScreen
      name="build"
    />
    <TabsScreen
      name="server"
    />
  </Tabs>
</View>
`;

exports[`TabsContainer renders with current extension ID 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <Tabs
    screenOptions={
      {
        "headerShown": false,
        "tabBarActiveTintColor": "#007AFF",
        "tabBarInactiveTintColor": "#8E8E93",
        "tabBarStyle": {
          "backgroundColor": "#F2F2F7",
          "borderTopColor": "#C6C6C8",
        },
      }
    }
  >
    <TabsScreen
      name="index"
    />
    <TabsScreen
      name="extensions/[id]"
    />
    <TabsScreen
      name="build"
    />
    <TabsScreen
      name="server"
    />
  </Tabs>
</View>
`;

exports[`TabsContainer renders with default state 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <Tabs
    screenOptions={
      {
        "headerShown": false,
        "tabBarActiveTintColor": "#007AFF",
        "tabBarInactiveTintColor": "#8E8E93",
        "tabBarStyle": {
          "backgroundColor": "#F2F2F7",
          "borderTopColor": "#C6C6C8",
        },
      }
    }
  >
    <TabsScreen
      name="index"
    />
    <TabsScreen
      name="extensions/[id]"
    />
    <TabsScreen
      name="build"
    />
    <TabsScreen
      name="server"
    />
  </Tabs>
</View>
`;
